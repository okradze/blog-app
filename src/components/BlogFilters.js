import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { setTitleFilter, searchByAuthor, searchByTitle } from '../actions/filters';

export const BlogFilters = ({ setTitleFilter, searchByTitle, searchByAuthor, filters }) => {
	const onTitleChange = e => {
		setTitleFilter(e.target.value);
	};

	const onSortChange = e => {
		if (e.target.value === 'title') {
			searchByTitle();
		} else if (e.target.value === 'author') {
			searchByAuthor();
		}
	};

	useEffect(() => () => setTitleFilter(''), []);

	return (
		<div className="filters">
			<input
				type="text"
				className="input filters__input"
				placeholder="Search..."
				value={filters.title}
				onChange={onTitleChange}
			/>
			<select className="input" value={filters.searchBy} onChange={onSortChange}>
				<option value="title">Title</option>
				<option value="author">Author</option>
			</select>
		</div>
	);
};

const mapStateToProps = state => ({
	filters: state.filters,
});

const mapDispatchToProps = dispatch => ({
	setTitleFilter: title => dispatch(setTitleFilter(title)),
	searchByAuthor: () => dispatch(searchByAuthor()),
	searchByTitle: () => dispatch(searchByTitle()),
});

export default connect(mapStateToProps, mapDispatchToProps)(BlogFilters);
