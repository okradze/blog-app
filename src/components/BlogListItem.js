import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import 'intersection-observer';

export const BlogListItem = ({ title, author, id, to, createdAt, url, photoURL }) => {
	const [ visibleCaption, setVisibleCaption ] = useState(true);

	return (
		<article className="blog-item">
			<div className="blog-info">
				<img src={photoURL} className="blog-info__photo" alt={author} />
				<div className="blog-info__content">
					<p className="blog-info__author">{author}</p>
					<time dateTime={new Date(createdAt)} className="blog-info__date">
						{moment(createdAt).format('MMM Do, YYYY')}
					</time>
				</div>
			</div>

			<figure className="blog-item__thumbnail">
				{visibleCaption && <div className="blog-item__thumbnail-no" />}
				<img onLoad={() => setVisibleCaption()} src={url} alt={title} className="blog-item__thumbnail-img" />
			</figure>

			<div className="blog-item__content">
				<h2 className="blog-item__title">{title}</h2>
				{to === 'edit' && (
					<Link className="read-more" to={`${to}/${id}`}>
						Edit Blog
					</Link>
				)}
				<Link className="read-more" to={`read/${id}`}>
					Read more...
				</Link>
			</div>
		</article>
	);
};
export default BlogListItem;
