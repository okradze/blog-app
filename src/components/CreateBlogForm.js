import React, { useState } from 'react';
import { connect } from 'react-redux';
import xss from 'xss';
import Quill from './Quill';
import loader from '../images/loader.gif';

export const CreateBlogForm = ({
	onSubmit,
	title,
	body,
	createdAt,
	submitDisabled,
	submitLoading,
	isEdit,
	removeDisabled,
	removeLoading,
	onRemove,
	auth,
}) => {
	const [ blogTitle, setTitle ] = useState(title || '');
	const [ blogBody, setBody ] = useState(body || '');
	const [ blogCreatedAt ] = useState(createdAt || new Date().getTime());
	const [ file, setFile ] = useState(null);
	const [ error, setError ] = useState('');

	const onTitleChange = e => {
		setTitle(e.target.value);
	};
	const onBodyChange = value => {
		setBody(value);
	};
	const onSave = e => {
		e.preventDefault();
		if (!blogTitle || !blogBody) {
			setError('Please provide blog title and body');
		} else {
			setError('');
			onSubmit(
				{
					title: xss(blogTitle),
					body: xss(blogBody),
					createdAt: blogCreatedAt,
					author: xss(auth.displayName),
					photoURL: xss(auth.photoURL),
					file,
				},
				auth.uid,
			);
		}
	};

	const onFileChange = e => {
		if (e.target.files[0]) {
			setFile(e.target.files[0]);
		}
	};

	return (
		<form className="create-form" onSubmit={onSave}>
			{error && (
				<div className="alert create-form__alert" role="alert">
					<p>{error}</p>
				</div>
			)}
			<input
				type="text"
				className="input create-form__input"
				placeholder="Blog title"
				value={blogTitle}
				onChange={onTitleChange}
			/>

			<input type="file" className="input" onChange={onFileChange} />

			<Quill body={blogBody} setBody={onBodyChange} />
			<button onClick={onSave} type="button" className="button create-form__button" disabled={submitDisabled}>
				{submitLoading && <img className="w-4 h-4 mr-2" src={loader} alt="loader" />} Save Blog
			</button>
			{isEdit && (
				<button type="button" className="button button--grey" onClick={onRemove} disabled={removeDisabled}>
					{removeLoading && <img className="center small-loader" src={loader} alt="loader" />} Remove
				</button>
			)}
		</form>
	);
};

const mapStateToProps = state => ({
	auth: state.auth,
});

export default connect(mapStateToProps)(CreateBlogForm);
