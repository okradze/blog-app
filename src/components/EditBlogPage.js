import React from 'react';

export const EditBlogPage = (props) => (
    <div>Edit blog {props.match.params.id}</div>
);

export default EditBlogPage;