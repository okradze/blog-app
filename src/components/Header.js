import React from 'react';

export const Header = () => (
    <header>
        <h1>Blog</h1>
        <button>Logout</button>
    </header>
);

export default Header;