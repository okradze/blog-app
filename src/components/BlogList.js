import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import 'intersection-observer';
import db from '../firebase/firebase';
import BlogListItem from './BlogListItem';
import loader from '../images/loader.gif';
import selectBlogs from '../selectors/blogs';

let startDoc = null;

const getBlogs = docs => {
	const blogs = [];
	docs.forEach(doc => {
		blogs.push({
			id: doc.id,
			...doc.data(),
		});
	});
	startDoc = docs[docs.length - 1];
	return blogs;
};

const firstFetchBlogs = async uid => {
	const data = await db.collection('users').doc(uid).collection('blogs').orderBy('createdAt', 'desc').limit(6).get();
	return getBlogs(data.docs);
};

const fetchBlogs = async uid => {
	const data = await db
		.collection('users')
		.doc(uid)
		.collection('blogs')
		.orderBy('createdAt', 'desc')
		.startAfter(startDoc)
		.limit(6)
		.get();
	return getBlogs(data.docs);
};

export const BlogList = ({ uid, filters }) => {
	const [ blogs, setBlogs ] = useState([]);
	const [ loading, setLoading ] = useState(true);
	const [ fetching, setFetching ] = useState(false);

	const scrollElement = React.useRef();

	const paginateBlogs = () => {
		setFetching(true);
		fetchBlogs(uid)
			.then(newBlogs => {
				const data = JSON.parse(sessionStorage.getItem('blogs'));
				setBlogs([ ...data, ...newBlogs ]);
				sessionStorage.setItem('blogs', JSON.stringify([ ...data, ...newBlogs ]));
				setFetching(false);
			})
			.catch(() => {
				setFetching(false);
			});
	};

	useEffect(() => {
		let didCancel = false;
		firstFetchBlogs(uid).then(newBlogs => {
			if (!didCancel) {
				setBlogs(newBlogs);
				sessionStorage.setItem('blogs', JSON.stringify(newBlogs));
				setLoading(false);

				const io = new IntersectionObserver(entries => {
					const ratio = entries[0].intersectionRatio;
					if (ratio > 0) {
						setFetching(true);
						paginateBlogs();
					}
				});
				io.observe(scrollElement.current);
			}
		});
		return () => {
			didCancel = true;
		};
	}, []);

	const filteredBlogs = selectBlogs(blogs, filters);

	return (
		<React.Fragment>
			{loading ? (
				<img className="center small-loader" src={loader} alt="Loader" />
			) : (
				<div className="blog-list">
					{filteredBlogs.map(blog => <BlogListItem to="edit" key={blog.id} {...blog} />)}
					{fetching && <img className="small-loader" src={loader} alt="Loader" />}
					<div style={{ minHeight: '1px', minWidth: '1px' }} ref={scrollElement} />
				</div>
			)}
		</React.Fragment>
	);
};

const mapStateToProps = state => ({
	uid: state.auth.uid,
	filters: state.filters,
});

export default connect(mapStateToProps)(BlogList);
