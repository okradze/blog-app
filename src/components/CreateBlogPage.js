import React, { useState } from 'react';
import CreateBlogForm from './CreateBlogForm';
import { startCreateBlog } from '../actions/blogs';

export const CreateBlogPage = ({ history }) => {
	const [ submitLoading, setSubmitLoading ] = useState(false);
	const [ submitDisabled, setSubmitDisabled ] = useState(false);

	const onSubmit = (blog, uid) => {
		setSubmitLoading(true);
		setSubmitDisabled(true);
		startCreateBlog(blog, uid).then(() => {
			history.push('/dashboard');
		});
	};
	return (
		<React.Fragment>
			<div className="container mb-auto">
				<CreateBlogForm submitDisabled={submitDisabled} submitLoading={submitLoading} onSubmit={onSubmit} />
			</div>
		</React.Fragment>
	);
};

export default CreateBlogPage;
