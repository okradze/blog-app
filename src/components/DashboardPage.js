import React from 'react';
import BlogFilters from './BlogFilters';
import BlogList from './BlogList';

export const DashboardPage = () => (
    <div className="dashboard-page container">
        <BlogFilters />
        <BlogList />
    </div>
);

export default DashboardPage;