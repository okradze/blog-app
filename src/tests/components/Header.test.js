import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';

let startLogin, startLogout, photoURL, uid, loggedInWrapper, loggedOutWrapper;

beforeEach(() => {
	startLogin = jest.fn();
	startLogout = jest.fn();
	photoURL = 'url';
	uid = 'abc123';
	loggedInWrapper = shallow(
		<Header startLogin={startLogin} startLogout={startLogout} photoURL={photoURL} uid={uid} />,
	);
	loggedOutWrapper = shallow(<Header startLogin={startLogin} />);
});

test('should render Header correctly when logged in', () => {
	expect(loggedInWrapper).toMatchSnapshot();
});

test('should render Header correctly when logged out', () => {
	expect(loggedOutWrapper).toMatchSnapshot();
});

test('should call startLogout on button click', () => {
	loggedInWrapper.find('button').simulate('click');
	expect(startLogout).toHaveBeenCalled();
});

test('should call startLogin on button click', () => {
	loggedOutWrapper.find('button').at(0).simulate('click');
	loggedOutWrapper.find('button').at(1).simulate('click');
	expect(startLogin).toHaveBeenCalledTimes(2);
});
