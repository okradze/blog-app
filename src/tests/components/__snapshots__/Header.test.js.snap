// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly when logged in 1`] = `
<header
  className="header"
>
  <div
    className="container header__content"
  >
    <div
      className="mobile-toggle"
    >
      <label
        className="mobile-toggle__label"
        htmlFor="nav-toggle"
      >
        <img
          alt="Hamburger logo for mobile devices"
          className="mobile-toggle__hamburger"
          src={Object {}}
        />
      </label>
      <input
        className="mobile-toggle__checkbox"
        id="nav-toggle"
        type="checkbox"
      />
      <nav
        className="mobile-nav"
      >
        <Link
          className="mobile-nav__link"
          replace={false}
          to="/dashboard"
        >
          Dashboard
        </Link>
        <Link
          className="mobile-nav__link"
          replace={false}
          to="/create"
        >
          Create
        </Link>
        <Link
          className="mobile-nav__link"
          replace={false}
          to="/read"
        >
          Blogs
        </Link>
      </nav>
    </div>
    <div
      className="desktop-toggle"
    >
      <nav
        className="nav"
      >
        <Link
          className="nav__link"
          replace={false}
          to="/dashboard"
        >
          Dashboard
        </Link>
        <Link
          className="nav__link"
          replace={false}
          to="/create"
        >
          Create
        </Link>
        <Link
          className="nav__link"
          replace={false}
          to="/read"
        >
          Blogs
        </Link>
      </nav>
      <div
        className="header__logo"
      >
        <Link
          className="header__logo-text"
          replace={false}
          to="/dashboard"
        >
          <h1>
            Blog App
          </h1>
        </Link>
      </div>
      <div
        className="header__user"
      >
        <img
          alt="User"
          className="header__img"
          src="url"
        />
        <button
          className="button"
          onClick={[MockFunction]}
          type="button"
        >
          Log Out
        </button>
      </div>
    </div>
  </div>
</header>
`;

exports[`should render Header correctly when logged out 1`] = `
<header
  className="header"
>
  <div
    className="container header__content"
  >
    <div
      className="mobile-toggle"
    >
      <label
        className="mobile-toggle__label"
        htmlFor="nav-toggle"
      >
        <img
          alt="Hamburger logo for mobile devices"
          className="mobile-toggle__hamburger"
          src={Object {}}
        />
      </label>
      <input
        className="mobile-toggle__checkbox"
        id="nav-toggle"
        type="checkbox"
      />
      <nav
        className="mobile-nav"
      >
        <button
          className="mobile-nav__link"
          onClick={[MockFunction]}
          type="button"
        >
          Login
        </button>
      </nav>
    </div>
    <div
      className="desktop-toggle"
    >
      <div
        className="header__logo"
      >
        <Link
          className="header__logo-text"
          replace={false}
          to="/dashboard"
        >
          <h1>
            Blog App
          </h1>
        </Link>
      </div>
      <div
        className="header__user"
      >
        <button
          className="button"
          onClick={[MockFunction]}
          type="button"
        >
          Log In
        </button>
      </div>
    </div>
  </div>
</header>
`;
